File: experiment/backends/android/valkey/ValkeyContextManager.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package dev.dazzle.experiment

import dev.dazzle.sdk.DazzleServer
import java.io.BufferedInputStream
import java.io.BufferedWriter
import java.io.OutputStreamWriter
import java.net.Socket
import kotlin.math.max

/**
 * Stock-Valkey baseline that speaks RESP2 over TCP loopback.
 *
 * This backend is the "Valkey as you'd use it today" reference point in
 * the paper's comparison. Where [DazzleContextManager] dispatches every
 * command through the in-process pipe → ring → io_uring pipeline, this
 * one opens a real TCP socket to `127.0.0.1:<port>` and writes RESP2
 * frames on the wire — exactly what any app using a Jedis/Lettuce
 * client against a co-located Valkey server would pay per command.
 *
 * The server it talks to is still the same embedded Dazzle-patched
 * Valkey build (so command semantics match byte-for-byte). What varies
 * is the transport path:
 *
 *   Dazzle : app thread → self-pipe → server ae loop
 *   Valkey : app thread → kernel TCP → epoll → server ae loop
 *
 * Both backends store data in the exact same keys with the exact same
 * commands, so `buildContextBlock` produces byte-equivalent prompts;
 * the only variable across runs is the transport.
 */
class ValkeyContextManager(
    private val host: String = "127.0.0.1",
    private val port: Int = DazzleServer.getPort().takeIf { it > 0 } ?: 6380,
) : StorageBackend {

    override val backendName: String = "Valkey"

    private val socket: Socket = Socket(host, port).apply {
        tcpNoDelay = true
        soTimeout  = 10_000
    }
    private val writer = BufferedWriter(OutputStreamWriter(socket.getOutputStream(), Charsets.UTF_8))
    private val reader = BufferedInputStream(socket.getInputStream())

    // ── RESP encode / decode ─────────────────────────────────────────────

    /** Send one command as a RESP2 array and block until the reply parses. */
    private fun send(vararg args: String): Reply {
        val sb = StringBuilder()
        sb.append('*').append(args.size).append("\r\n")
        for (a in args) {
            sb.append('$').append(a.toByteArray(Charsets.UTF_8).size).append("\r\n")
            sb.append(a).append("\r\n")
        }
        writer.write(sb.toString())
        writer.flush()
        return readReply()
    }

    private fun readReply(): Reply {
        val marker = reader.read()
        if (marker < 0) throw IllegalStateException("socket closed")
        return when (marker.toChar()) {
            '+' -> Reply.Simple(readLine())
            '-' -> Reply.Error(readLine())
            ':' -> Reply.Integer(readLine().toLong())
            '$' -> {
                val n = readLine().toInt()
                if (n < 0) Reply.Bulk(null)
                else Reply.Bulk(readExact(n).also { readCrlf() })
            }
            '*' -> {
                val n = readLine().toInt()
                if (n < 0) Reply.Array(null)
                else Reply.Array(List(n) { readReply() })
            }
            else -> throw IllegalStateException("unknown RESP marker '${marker.toChar()}'")
        }
    }

    private fun readLine(): String {
        val out = StringBuilder()
        while (true) {
            val c = reader.read()
            if (c < 0) throw IllegalStateException("socket closed mid-line")
            if (c == '\r'.code) { reader.read(); return out.toString() }
            out.append(c.toChar())
        }
    }

    private fun readExact(n: Int): String {
        val buf = ByteArray(n)
        var off = 0
        while (off < n) {
            val r = reader.read(buf, off, n - off)
            if (r < 0) throw IllegalStateException("socket closed mid-bulk")
            off += r
        }
        return String(buf, Charsets.UTF_8)
    }

    private fun readCrlf() { reader.read(); reader.read() }

    private sealed class Reply {
        data class Simple(val value: String) : Reply()
        data class Error(val message: String) : Reply()
        data class Integer(val value: Long) : Reply()
        data class Bulk(val value: String?) : Reply()
        data class Array(val items: List<Reply>?) : Reply()
    }

    private fun Reply.asBulkOrNull(): String? = when (this) {
        is Reply.Bulk    -> value
        is Reply.Simple  -> value
        is Reply.Integer -> value.toString()
        else             -> null
    }

    private fun Reply.asArrayOfBulks(): List<String?> = when (this) {
        is Reply.Array   -> items?.map { it.asBulkOrNull() }.orEmpty()
        else             -> emptyList()
    }

    // ── StorageBackend ────────────────────────────────────────────────────

    override fun ingest(reading: SensorReading) {
        send(
            "XADD", "sensor:readings", "MAXLEN", "~", "200", "*",
            "temp",      reading.tempC.toString(),
            "humidity",  reading.humidity.toString(),
            "minute",    reading.minute.toString(),
            "anomalous", if (reading.anomalous) "1" else "0",
        )
        send("HINCRBYFLOAT", "sensor:stats", "temp_sum", reading.tempC.toString())
        send("HINCRBY",      "sensor:stats", "count",    "1")
        send("HSET",         "sensor:stats", "latest_temp",   reading.tempC.toString(),
                                              "latest_minute", reading.minute.toString())

        val curMin = send("HGET", "sensor:stats", "min_temp").asBulkOrNull()?.toDoubleOrNull()
        if (curMin == null || reading.tempC < curMin) {
            send("HSET", "sensor:stats", "min_temp", reading.tempC.toString())
        }
        val curMax = send("HGET", "sensor:stats", "max_temp").asBulkOrNull()?.toDoubleOrNull()
        if (curMax == null || reading.tempC > curMax) {
            send("HSET", "sensor:stats", "max_temp", reading.tempC.toString())
        }

        if (reading.anomalous) {
            send("ZADD", "sensor:anomalies", reading.minute.toString(), reading.minute.toString())
            send("HINCRBY", "sensor:stats", "anomaly_count", "1")
        }
    }

    override fun flush() {
        send("DEL", "sensor:readings", "sensor:stats", "sensor:anomalies", "agent:decisions")
        for (i in 0..9) send("DEL", "agent:checkpoint:$i")
    }

    override fun buildContextBlock(currentMinute: Int, windowMinutes: Int): String {
        val lines = mutableListOf<String>()

        // Last 10 entries — newest first
        val entries = send("XREVRANGE", "sensor:readings", "+", "-", "COUNT", "10")
        val temps = mutableListOf<Double>()
        (entries as? Reply.Array)?.items?.forEach { entry ->
            // entry is [id, [field, value, field, value, ...]]
            val pair = (entry as? Reply.Array)?.items ?: return@forEach
            val fields = (pair.getOrNull(1) as? Reply.Array)?.items ?: return@forEach
            var i = 0
            while (i < fields.size - 1) {
                if ((fields[i] as? Reply.Bulk)?.value == "temp") {
                    (fields[i + 1] as? Reply.Bulk)?.value?.toDoubleOrNull()?.let { temps.add(it) }
                    break
                }
                i += 2
            }
        }
        val recentTemps = temps.reversed()
        if (recentTemps.isNotEmpty()) {
            val formatted = recentTemps.joinToString(", ") { String.format("%.1f", it) }
            lines += "Last ${recentTemps.size} temperatures (oldest→newest, °C): $formatted"
            lines += "Recent trend: ${computeTrend(recentTemps)}"
        }

        // Aggregate stats via a single HMGET
        val statsReply = send("HMGET", "sensor:stats", "count", "temp_sum", "min_temp", "max_temp", "anomaly_count")
        val vals = statsReply.asArrayOfBulks()
        val count = vals.getOrNull(0)?.toIntOrNull()
        if (count != null && count > 0) {
            val sum    = vals.getOrNull(1)?.toDoubleOrNull() ?: 0.0
            val minT   = vals.getOrNull(2)?.toDoubleOrNull() ?: 0.0
            val maxT   = vals.getOrNull(3)?.toDoubleOrNull() ?: 0.0
            val anomCt = vals.getOrNull(4)?.toIntOrNull()    ?: 0
            lines += "Aggregate over $count readings: " +
                "avg=${String.format("%.1f", sum / count)}°C, " +
                "min=${String.format("%.1f", minT)}°C, " +
                "max=${String.format("%.1f", maxT)}°C"
            lines += "Total anomalies detected so far: $anomCt"
        }

        val windowStart = max(0, currentMinute - windowMinutes)
        val anomaliesInWindow = send(
            "ZRANGEBYSCORE", "sensor:anomalies",
            windowStart.toString(), currentMinute.toString()
        ).asArrayOfBulks().filterNotNull()
        if (anomaliesInWindow.isEmpty()) {
            lines += "No anomalies in the last $windowMinutes minutes."
        } else {
            lines += "Anomalous time indices in the last $windowMinutes minutes " +
                "(minute numbers, not temperatures): [${anomaliesInWindow.joinToString(", ")}]"
        }

        return lines.joinToString("\n")
    }

    override fun buildSynthesisContext(): String {
        val lines = mutableListOf<String>()
        val vals = send(
            "HMGET", "sensor:stats",
            "count", "temp_sum", "min_temp", "max_temp", "anomaly_count"
        ).asArrayOfBulks()
        val count = vals.getOrNull(0)?.toIntOrNull()
        if (count != null && count > 0) {
            val sum    = vals.getOrNull(1)?.toDoubleOrNull() ?: 0.0
            val minT   = vals.getOrNull(2)?.toDoubleOrNull() ?: 0.0
            val maxT   = vals.getOrNull(3)?.toDoubleOrNull() ?: 0.0
            val anomCt = vals.getOrNull(4)?.toIntOrNull()    ?: 0
            lines += "=== Full Session Stats ==="
            lines += "Total readings: $count"
            lines += "Temperature range: ${String.format("%.1f", minT)}°C to " +
                "${String.format("%.1f", maxT)}°C " +
                "(avg ${String.format("%.1f", sum / count)}°C)"
            lines += "Total anomalies detected: $anomCt"
        }

        val allAnomalies = send(
            "ZRANGEBYSCORE", "sensor:anomalies", "0", "99999"
        ).asArrayOfBulks().filterNotNull()
        if (allAnomalies.isNotEmpty()) {
            lines += "Anomalous time indices (minute numbers, not temperatures): " +
                "[${allAnomalies.joinToString(", ")}]"
        }

        val decisions = send("LRANGE", "agent:decisions", "0", "-1").asArrayOfBulks().filterNotNull()
        if (decisions.isNotEmpty()) {
            lines += "=== Monitoring Agent Decisions ==="
            decisions.forEachIndexed { i, d -> lines += "  Checkpoint ${i + 1}: $d" }
        }

        return lines.joinToString("\n")
    }

    override fun storeCheckpointDecision(
        index: Int,
        minute: Int,
        anomalyDetected: Boolean,
        severity: String,
        trend: String,
    ) {
        val decision = "anomaly=${if (anomalyDetected) "yes" else "no"} " +
            "severity=$severity trend=$trend"

        send(
            "HSET", "agent:checkpoint:$index",
            "minute",   minute.toString(),
            "anomaly",  if (anomalyDetected) "1" else "0",
            "severity", severity,
            "trend",    trend,
        )
        send("RPUSH", "agent:decisions", decision)
    }

    override fun measureRetrievalLatency(currentMinute: Int): Double {
        val start = System.nanoTime()
        buildContextBlock(currentMinute)
        val end = System.nanoTime()
        return (end - start) / 1_000.0
    }

    private fun computeTrend(temps: List<Double>): String {
        if (temps.size < 2) return "stable"
        val n = temps.size
        val meanX = (n - 1) / 2.0
        val meanY = temps.average()
        val num = (0 until n).sumOf { (it - meanX) * (temps[it] - meanY) }
        val den = (0 until n).sumOf { (it - meanX) * (it - meanX) }
        val slope = if (den != 0.0) num / den else 0.0
        return when {
            slope >  0.15 -> "increasing"
            slope < -0.15 -> "decreasing"
            else          -> "stable"
        }
    }

    // ── Footprint accounting ──────────────────────────────────────────────

    override val backendSizeMethod: String = "valkey:used_memory_dataset"

    /**
     * Issues an `INFO memory` over the same RESP socket and parses
     * the `used_memory_dataset` line. Stays loyal to the "as you'd
     * use Valkey in production" framing of this backend instead of
     * borrowing the in-process Dazzle SDK.
     */
    override fun backendSizeBytes(): Long {
        val raw = send("INFO", "memory").asBulkOrNull() ?: return -1L
        for (line in raw.lineSequence()) {
            val trimmed = line.trim()
            val colon = trimmed.indexOf(':')
            if (colon <= 0) continue
            val key = trimmed.substring(0, colon)
            if (key == "used_memory_dataset" || key == "used_memory") {
                val v = trimmed.substring(colon + 1).toLongOrNull()
                if (v != null && key == "used_memory_dataset") return v
                if (v != null && key == "used_memory") return v // fallback
            }
        }
        return -1L
    }
}

File: experiment/backends/ios/valkey/ValkeyContextManager.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

/// Stock-Valkey baseline that speaks RESP2 over TCP loopback.
///
/// Companion to `DazzleContextManager`: same data layout, same commands,
/// but the transport is a real TCP socket to `127.0.0.1:<port>` — the
/// path any app using a Jedis/Lettuce-style Valkey client would take
/// today. The server it talks to is still the same embedded Dazzle
/// build, so command semantics match byte-for-byte; the only variable
/// across runs is the transport.
///
///   Dazzle : app thread → self-pipe → server ae loop
///   Valkey : app thread → kernel TCP → kqueue → server ae loop
///
/// Uses one persistent socket protected by an NSLock so it mirrors the
/// cost model a real mobile client would have (no ephemeral-port churn).
final class ValkeyContextManager: StorageBackend {

    let backendName: String = "Valkey"

    private var input:  InputStream!
    private var output: OutputStream!
    private let lock = NSLock()

    // Pre-allocated 64 KB read buffer, grown on demand by `readBytes`.
    private var readBuf = [UInt8](repeating: 0, count: 64 * 1024)

    init(host: String = "127.0.0.1", port: Int = DazzleServer.shared.port) {
        var is_: InputStream?
        var os_: OutputStream?
        Stream.getStreamsToHost(withName: host, port: port, inputStream: &is_, outputStream: &os_)
        guard let i = is_, let o = os_ else {
            fatalError("ValkeyContextManager: could not open TCP socket to \(host):\(port)")
        }
        i.open(); o.open()
        self.input = i
        self.output = o
    }

    // ── RESP encode ──────────────────────────────────────────────────────

    @discardableResult
    private func send(_ args: String...) -> Reply {
        lock.lock(); defer { lock.unlock() }
        var payload = "*\(args.count)\r\n"
        for a in args {
            let bytes = a.utf8.count
            payload += "$\(bytes)\r\n\(a)\r\n"
        }
        let data = Array(payload.utf8)
        _ = data.withUnsafeBufferPointer { output.write($0.baseAddress!, maxLength: data.count) }
        return readReply()
    }

    // ── RESP decode ──────────────────────────────────────────────────────

    private func readReply() -> Reply {
        let marker = readByte()
        switch marker {
        case UInt8(ascii: "+"): return .simple(readLine())
        case UInt8(ascii: "-"): return .error(readLine())
        case UInt8(ascii: ":"): return .integer(Int64(readLine()) ?? 0)
        case UInt8(ascii: "$"):
            let n = Int(readLine()) ?? -1
            if n < 0 { return .bulk(nil) }
            let s = readExact(n)
            _ = readByte(); _ = readByte()   // CRLF
            return .bulk(s)
        case UInt8(ascii: "*"):
            let n = Int(readLine()) ?? -1
            if n < 0 { return .array(nil) }
            var items: [Reply] = []
            items.reserveCapacity(n)
            for _ in 0..<n { items.append(readReply()) }
            return .array(items)
        default:
            return .error("bad RESP marker \(marker)")
        }
    }

    private func readByte() -> UInt8 {
        var byte: UInt8 = 0
        let n = withUnsafeMutablePointer(to: &byte) { input.read($0, maxLength: 1) }
        precondition(n == 1, "socket EOF")
        return byte
    }

    private func readLine() -> String {
        var out: [UInt8] = []
        while true {
            let b = readByte()
            if b == 0x0D { _ = readByte(); return String(bytes: out, encoding: .utf8) ?? "" }
            out.append(b)
        }
    }

    private func readExact(_ count: Int) -> String {
        if count > readBuf.count { readBuf = [UInt8](repeating: 0, count: count) }
        var off = 0
        while off < count {
            let r = readBuf.withUnsafeMutableBufferPointer { input.read($0.baseAddress! + off, maxLength: count - off) }
            precondition(r > 0, "socket EOF")
            off += r
        }
        return String(bytes: readBuf[0..<count], encoding: .utf8) ?? ""
    }

    private enum Reply {
        case simple(String)
        case error(String)
        case integer(Int64)
        case bulk(String?)
        case array([Reply]?)

        var bulkOrNil: String? {
            switch self {
            case .bulk(let v):   return v
            case .simple(let v): return v
            case .integer(let v): return String(v)
            default:              return nil
            }
        }
        var arrayOfBulks: [String?] {
            if case .array(let items) = self { return (items ?? []).map { $0.bulkOrNil } }
            return []
        }
    }

    // ── StorageBackend ────────────────────────────────────────────────────

    func flush() {
        _ = send("DEL", "sensor:readings", "sensor:stats", "sensor:anomalies", "agent:decisions")
        for i in 0...9 { _ = send("DEL", "agent:checkpoint:\(i)") }
    }

    func ingest(_ reading: SensorReading) {
        _ = send(
            "XADD", "sensor:readings", "MAXLEN", "~", "200", "*",
            "temp",      String(reading.tempC),
            "humidity",  String(reading.humidity),
            "minute",    String(reading.minute),
            "anomalous", reading.anomalous ? "1" : "0"
        )
        _ = send("HINCRBYFLOAT", "sensor:stats", "temp_sum", String(reading.tempC))
        _ = send("HINCRBY",      "sensor:stats", "count",    "1")
        _ = send("HSET", "sensor:stats",
                 "latest_temp",   String(reading.tempC),
                 "latest_minute", String(reading.minute))

        if let curMin = send("HGET", "sensor:stats", "min_temp").bulkOrNil.flatMap(Double.init) {
            if reading.tempC < curMin {
                _ = send("HSET", "sensor:stats", "min_temp", String(reading.tempC))
            }
        } else {
            _ = send("HSET", "sensor:stats", "min_temp", String(reading.tempC))
        }
        if let curMax = send("HGET", "sensor:stats", "max_temp").bulkOrNil.flatMap(Double.init) {
            if reading.tempC > curMax {
                _ = send("HSET", "sensor:stats", "max_temp", String(reading.tempC))
            }
        } else {
            _ = send("HSET", "sensor:stats", "max_temp", String(reading.tempC))
        }

        if reading.anomalous {
            _ = send("ZADD", "sensor:anomalies",
                     String(reading.minute), String(reading.minute))
            _ = send("HINCRBY", "sensor:stats", "anomaly_count", "1")
        }
    }

    func buildContextBlock(currentMinute: Int, windowMinutes: Int) -> String {
        var lines: [String] = []

        // Last 10 entries — newest first
        var temps: [Double] = []
        if case .array(let items) = send("XREVRANGE", "sensor:readings", "+", "-", "COUNT", "10") {
            for entry in items ?? [] {
                guard case .array(let pair?) = entry, pair.count >= 2 else { continue }
                guard case .array(let fields?) = pair[1] else { continue }
                var i = 0
                while i < fields.count - 1 {
                    if case .bulk(let k?) = fields[i], k == "temp" {
                        if case .bulk(let v?) = fields[i + 1], let t = Double(v) { temps.append(t) }
                        break
                    }
                    i += 2
                }
            }
        }
        let recentTemps = Array(temps.reversed())
        if !recentTemps.isEmpty {
            let formatted = recentTemps.map { String(format: "%.1f", $0) }.joined(separator: ", ")
            lines.append("Last \(recentTemps.count) temperatures (oldest→newest, °C): \(formatted)")
            lines.append("Recent trend: \(computeTrend(recentTemps))")
        }

        let vals = send(
            "HMGET", "sensor:stats",
            "count", "temp_sum", "min_temp", "max_temp", "anomaly_count"
        ).arrayOfBulks
        if vals.count == 5, let count = vals[0].flatMap(Int.init), count > 0 {
            let sum    = vals[1].flatMap(Double.init) ?? 0
            let minT   = vals[2].flatMap(Double.init) ?? 0
            let maxT   = vals[3].flatMap(Double.init) ?? 0
            let anomCt = vals[4].flatMap(Int.init) ?? 0
            lines.append("Aggregate over \(count) readings: " +
                "avg=\(String(format: "%.1f", sum / Double(count)))°C, " +
                "min=\(String(format: "%.1f", minT))°C, " +
                "max=\(String(format: "%.1f", maxT))°C")
            lines.append("Total anomalies detected so far: \(anomCt)")
        }

        let windowStart = max(0, currentMinute - windowMinutes)
        let anomaliesInWindow = send(
            "ZRANGEBYSCORE", "sensor:anomalies",
            String(windowStart), String(currentMinute)
        ).arrayOfBulks.compactMap { $0 }
        if anomaliesInWindow.isEmpty {
            lines.append("No anomalies in the last \(windowMinutes) minutes.")
        } else {
            lines.append("Anomalous time indices in the last \(windowMinutes) minutes " +
                "(minute numbers, not temperatures): [\(anomaliesInWindow.joined(separator: ", "))]")
        }

        return lines.joined(separator: "\n")
    }

    func buildSynthesisContext() -> String {
        var lines: [String] = []
        let vals = send(
            "HMGET", "sensor:stats",
            "count", "temp_sum", "min_temp", "max_temp", "anomaly_count"
        ).arrayOfBulks
        if vals.count == 5, let count = vals[0].flatMap(Int.init), count > 0 {
            let sum    = vals[1].flatMap(Double.init) ?? 0
            let minT   = vals[2].flatMap(Double.init) ?? 0
            let maxT   = vals[3].flatMap(Double.init) ?? 0
            let anomCt = vals[4].flatMap(Int.init) ?? 0
            lines.append("=== Full Session Stats ===")
            lines.append("Total readings: \(count)")
            lines.append("Temperature range: \(String(format: "%.1f", minT))°C to " +
                "\(String(format: "%.1f", maxT))°C " +
                "(avg \(String(format: "%.1f", sum / Double(count)))°C)")
            lines.append("Total anomalies detected: \(anomCt)")
        }

        let allAnomalies = send(
            "ZRANGEBYSCORE", "sensor:anomalies", "0", "99999"
        ).arrayOfBulks.compactMap { $0 }
        if !allAnomalies.isEmpty {
            lines.append("Anomalous time indices (minute numbers, not temperatures): " +
                "[\(allAnomalies.joined(separator: ", "))]")
        }

        let decisions = send("LRANGE", "agent:decisions", "0", "-1").arrayOfBulks.compactMap { $0 }
        if !decisions.isEmpty {
            lines.append("=== Monitoring Agent Decisions ===")
            for (i, d) in decisions.enumerated() {
                lines.append("  Checkpoint \(i + 1): \(d)")
            }
        }

        return lines.joined(separator: "\n")
    }

    func storeCheckpointDecision(
        index: Int, minute: Int,
        anomalyDetected: Bool, severity: String, trend: String
    ) {
        let decision = "anomaly=\(anomalyDetected ? "yes" : "no") severity=\(severity) trend=\(trend)"
        _ = send(
            "HSET", "agent:checkpoint:\(index)",
            "minute",   String(minute),
            "anomaly",  anomalyDetected ? "1" : "0",
            "severity", severity,
            "trend",    trend
        )
        _ = send("RPUSH", "agent:decisions", decision)
    }

    func measureRetrievalLatency(currentMinute: Int) -> Double {
        let start = DispatchTime.now().uptimeNanoseconds
        _ = buildContextBlock(currentMinute: currentMinute, windowMinutes: 20)
        let end = DispatchTime.now().uptimeNanoseconds
        return Double(end - start) / 1_000.0
    }

    private func computeTrend(_ temps: [Double]) -> String {
        guard temps.count >= 2 else { return "stable" }
        let n = temps.count
        let meanX = Double(n - 1) / 2.0
        let meanY = temps.reduce(0.0, +) / Double(n)
        var num = 0.0, den = 0.0
        for i in 0..<n {
            let dx = Double(i) - meanX
            num += dx * (temps[i] - meanY)
            den += dx * dx
        }
        let slope = den != 0 ? num / den : 0
        if slope >  0.15 { return "increasing" }
        if slope < -0.15 { return "decreasing" }
        return "stable"
    }

    // ── Footprint accounting ──────────────────────────────────────────────
    //
    // We talk RESP over the same TCP socket every other command uses, so
    // the byte count we report is what a real Lettuce/Jedis-style mobile
    // client would observe — no in-process shortcut.

    var backendSizeMethod: String { "valkey:used_memory_dataset" }

    func backendSizeBytes() -> Int64 {
        guard let raw = send("INFO", "memory").bulkOrNil else { return -1 }
        return parseValkeyUsedMemoryDataset(raw)
    }

    func backendSizeBreakdown() -> [String: Int64]? {
        guard let raw = send("INFO", "memory").bulkOrNil else { return nil }
        return parseValkeyMemoryStats(raw)
    }
}
